### PR TITLE
added notes about JLIcon's design to README

### PR DESCRIPTION
**Pull request by telamonian**
## References

#7299, #7700, #7767

There's been a lot of recent discussion about JLIcon. In order to help clarify things, I've added some notes about the design of JLIcon to the README of ui-components.

## Code changes

None

## User-facing changes

None

## Backwards-incompatible changes

None

### CODE REVIEW

File: packages/ui-components/README.md [hunk 1]
```diff
@@ -43,3 +43,47 @@ The icons are organized into various categories in `./style/icons`, based on whe
...
+- every icon is a symbol
+  - each icon is defined exactly once, and is thereafter referred to via a unique symbol
+  - to use an icon `fooIcon` outside of the file in which it is defined, the pattern will be to import `fooIcon`
+  - this enables compile-time checking; helps to ensure that icons are specified correctly when used (as opposed to the old pattern of specifying icon via a `string` with name or className)
+- every icon is flexible
+  - can used in any context in which icons are used in jlab
+    - every icon can be passsed into Lumino
+    - every icon can create a DOM node
+    - every icon can create a React component
+- dyanmic lookup (for when absolutely needed)
```
Comment (telamonian) [inline packages/ui-components/README.md]:
```suggestion
- dynamic lookup (for when absolutely needed)
```

**Comment by jtpio** [comment]
This should also fix https://github.com/jupyterlab/jupyterlab/issues/7142?

File: packages/ui-components/README.md [hunk 2]
```diff
@@ -43,3 +43,47 @@ The icons are organized into various categories in `./style/icons`, based on whe
...
+  - each icon is defined exactly once, and is thereafter referred to via a unique symbol
+  - to use an icon `fooIcon` outside of the file in which it is defined, the pattern will be to import `fooIcon`
+  - this enables compile-time checking; helps to ensure that icons are specified correctly when used (as opposed to the old pattern of specifying icon via a `string` with name or className)
+- every icon is flexible
+  - can used in any context in which icons are used in jlab
+    - every icon can be passsed into Lumino
+    - every icon can create a DOM node
+    - every icon can create a React component
+- dynamic lookup (for when absolutely needed)
+  - Use dynamic lookup for the few cases in which an icon can only be specified as a string (such as in json schema files)
```
Comment (jasongrout) [inline packages/ui-components/README.md]: So the string ids need to be unique? So we probably want to encourage namespacing the strings with the package name, like `'ui-components:myicon'`?

**Comment by telamonian** [comment]
@jtpio These particular docs probably aren't going to be of much interest/use to extension devs (mostly just notes on the conceptual design of JLIcon). I am also working on some dev docs, though

File: packages/ui-components/README.md [hunk 3]
```diff
@@ -43,3 +43,47 @@ The icons are organized into various categories in `./style/icons`, based on whe
...
+
+- the one true icon
+  - create a single, canonical, simple (as possible) way to set up and use icons in jlab
+- every icon is a symbol
+  - each icon is defined exactly once, and is thereafter referred to via a unique symbol
+  - to use an icon `fooIcon` outside of the file in which it is defined, the pattern will be to import `fooIcon`
+  - this enables compile-time checking; helps to ensure that icons are specified correctly when used (as opposed to the old pattern of specifying icon via a `string` with name or className)
+- every icon is flexible
+  - can used in any context in which icons are used in jlab
+    - every icon can be passsed into Lumino
```
Comment (jtpio) [inline packages/ui-components/README.md]: Typo spotted:

```suggestion
    - every icon can be passed into Lumino
```

Comment (telamonian) [reply packages/ui-components/README.md]: Thanks! 👍

File: packages/ui-components/README.md [hunk 4]
```diff
@@ -43,3 +43,47 @@ The icons are organized into various categories in `./style/icons`, based on whe
...
+  - every defined icon can be used any number of times in any set of contexts
+  - where an icon is defined should not matter; all icons defined in core and extensions are reusable in any other extension
+
+Design goals for JLIcon (partially implemented):
+
+- remove all other ways of creating icons (though leave an escape hatch)
+  - need to deprecate, then later remove, `iconClass` from a number of interfaces
+- replacable
+  - all icons can be customized by replacing their svg dynamically during runtime
+    - currently, I'm leaning towards the idea that icon replacements should be an (optional) part of a theme
```
Comment (saulshanabrook) [inline packages/ui-components/README.md]: That sounds good! Previously, some icons used to be in the themes right? https://github.com/jupyterlab/jupyterlab/blob/master/packages/theme-light-extension/style/urls.css#L6

**Comment by saulshanabrook** [comment]
Thanks for these notes! This all looks good to me!

**Comment by blink1073** [comment]
This work was merged into #7767

**Pull request closed without merging**